# Creating Prediction Dataset

# 1. Loading the dataset
**Starting with importing libraries and reading the dataset.**

In [1]:
import re
import warnings

import pandas as pd

from collections import Counter

warnings.filterwarnings("ignore")

In [2]:
dataframe = pd.read_csv("Datasets/steam_games.csv")

In [3]:
dataframe.head()

url types   
0    https://store.steampowered.com/app/379720/DOOM/   app  \
1  https://store.steampowered.com/app/578080/PLAY...   app   
2  https://store.steampowered.com/app/637090/BATT...   app   
3    https://store.steampowered.com/app/221100/DayZ/   app   
4  https://store.steampowered.com/app/8500/EVE_On...   app   

                            name   
0                           DOOM  \
1  PLAYERUNKNOWN'S BATTLEGROUNDS   
2                     BATTLETECH   
3                           DayZ   
4                     EVE Online   

                                        desc_snippet   
0  Now includes all three premium DLC packs (Unto...  \
1  PLAYERUNKNOWN'S BATTLEGROUNDS is a battle roya...   
2  Take command of your own mercenary outfit of '...   
3  The post-soviet country of Chernarus is struck...   
4  EVE Online is a community-driven spaceship MMO...   

                                      recent_reviews   
0  Very Positive,(554),- 89% of the 554 user revi...  \
1  Mixed,(6,214),- 49% of the 6,214 user reviews ...   
2  Mixed,(166),- 54% of the 166 user reviews in t...   
3  Mixed,(932),- 57% of the 932 user reviews in t...   
4  Mixed,(287),- 54% of the 287 user reviews in t...   

                                         all_reviews  release_date   
0  Very Positive,(42,550),- 92% of the 42,550 use...  May 12, 2016  \
1  Mixed,(836,608),- 49% of the 836,608 user revi...  Dec 21, 2017   
2  Mostly Positive,(7,030),- 71% of the 7,030 use...  Apr 24, 2018   
3  Mixed,(167,115),- 61% of the 167,115 user revi...  Dec 13, 2018   
4  Mostly Positive,(11,481),- 74% of the 11,481 u...   May 6, 2003   

             developer                                publisher   
0          id Software    Bethesda Softworks,Bethesda Softworks  \
1     PUBG Corporation        PUBG Corporation,PUBG Corporation   
2  Harebrained Schemes  Paradox Interactive,Paradox Interactive   
3  Bohemia Interactive  Bohemia Interactive,Bohemia Interactive   
4                  CCP                                  CCP,CCP   

                                        popular_tags   
0  FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
1  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
2  Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
3  Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
4  Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                        game_details   
0  Single-player,Multi-player,Co-op,Steam Achieve...  \
1             Multi-player,Online Multi-Player,Stats   
2  Single-player,Multi-player,Online Multi-Player...   
3  Multi-player,Online Multi-Player,Steam Worksho...   
4  Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                           languages  achievements   
0  English,French,Italian,German,Spanish - Spain,...          54.0  \
1  English,Korean,Simplified Chinese,French,Germa...          37.0   
2                      English,French,German,Russian         128.0   
3  English,French,Italian,German,Spanish - Spain,...           NaN   
4                      English,German,Russian,French           NaN   

                                               genre   
0                                             Action  \
1             Action,Adventure,Massively Multiplayer   
2                          Action,Adventure,Strategy   
3             Action,Adventure,Massively Multiplayer   
4  Action,Free to Play,Massively Multiplayer,RPG,...   

                                    game_description   
0   About This Game Developed by id software, the...  \
1   About This Game  PLAYERUNKNOWN'S BATTLEGROUND...   
2   About This Game  From original BATTLETECH/Mec...   
3   About This Game The post-soviet country of Ch...   
4                                  About This Game     

                                      mature_content   
0                                                NaN  \
1   Mature Content Description  The developers de...   
2                       

# 2. Cleaning the dataset
**I get the idea how data looks now. Next step is to check for missing values in dataset and drop rows with missing one.**

**2. Irrelevant columns**

In [4]:
dataframe_clean = dataframe.drop(["url", "types", "desc_snippet", "recent_reviews", "achievements",
                                      "game_description", "mature_content", "minimum_requirements",
                                     "recommended_requirements", "original_price", "discount_price"], axis=1)
dataframe_clean.tail()

name all_reviews   
40828  Rocksmith® 2014 Edition – Remastered – Sabaton...         NaN  \
40829  Rocksmith® 2014 Edition – Remastered – Stone T...         NaN   
40830  Fantasy Grounds - Quests of Doom 4: A Midnight...         NaN   
40831                       Mega Man X5 Sound Collection         NaN   
40832                                   Stories In Stone         NaN   

       release_date                developer                        publisher   
40828  Feb 12, 2019  Ubisoft - San Francisco                              NaN  \
40829   Feb 5, 2019  Ubisoft - San Francisco                              NaN   
40830  Jul 31, 2018      SmiteWorks USA, LLC                              NaN   
40831  Jul 24, 2018          CAPCOM CO., LTD  CAPCOM CO., LTD,CAPCOM CO., LTD   
40832   Aug 8, 2018      16 Bit Psych,Kyle B        Self-Publish,Self-Publish   

                                            popular_tags   
40828                                  Casual,Simulation  \
40829                                  Casual,Simulation   
40830  RPG,Indie,Strategy,Software,Turn-Based,Fantasy...   
40831                                             Action   
40832                                      RPG,Adventure   

                                            game_details   
40828  Single-player,Shared/Split Screen,Downloadable...  \
40829  Single-player,Shared/Split Screen,Downloadable...   
40830  Multi-player,Co-op,Cross-Platform Multiplayer,...   
40831  Single-player,Downloadable Content,Steam Achie...   
40832  Single-player,Partial Controller Support,Profi...   

                                               languages               genre  
40828  English,German,French,Italian,Spanish - Spain,...   Casual,Simulation  
40829  English,German,French,Italian,Spanish - Spain,...   Casual,Simulation  
40830                                            English  Indie,RPG,Strategy  
40831  English,French,Italian,German,Spanish - Spain,...              Action  
40832                                            English       Adventure,RPG

In [5]:
dataframe_clean.loc[40833] = [
    "Resident Evil 4",
    "Mixed,(16),- 54% of the 16 user reviews",
    "Mar 24, 2023",
    "CAPCOM CO., LTD",
    "CAPCOM CO., LTD",
    "Action,Horror,Survival Horror,Third-Person Shooter,Zombies,Survival,Singleplayer,Story Rich,Dark,Realistic,Cinematic,Resource Management,3D,Psychological Horror,Adventure,Gore,Remake,Third Person,Action-Adventure,Violent",
    "Single-player,Steam Achievements,Full controller support,Steam Trading Cards,Steam Cloud",
    "English,French,Italian,German,Spanish",
    "Action,Adventure"
]

In [6]:
dataframe_clean.tail()

name   
40829  Rocksmith® 2014 Edition – Remastered – Stone T...  \
40830  Fantasy Grounds - Quests of Doom 4: A Midnight...   
40831                       Mega Man X5 Sound Collection   
40832                                   Stories In Stone   
40833                                    Resident Evil 4   

                                   all_reviews  release_date   
40829                                      NaN   Feb 5, 2019  \
40830                                      NaN  Jul 31, 2018   
40831                                      NaN  Jul 24, 2018   
40832                                      NaN   Aug 8, 2018   
40833  Mixed,(16),- 54% of the 16 user reviews  Mar 24, 2023   

                     developer                        publisher   
40829  Ubisoft - San Francisco                              NaN  \
40830      SmiteWorks USA, LLC                              NaN   
40831          CAPCOM CO., LTD  CAPCOM CO., LTD,CAPCOM CO., LTD   
40832      16 Bit Psych,Kyle B        Self-Publish,Self-Publish   
40833          CAPCOM CO., LTD                  CAPCOM CO., LTD   

                                            popular_tags   
40829                                  Casual,Simulation  \
40830  RPG,Indie,Strategy,Software,Turn-Based,Fantasy...   
40831                                             Action   
40832                                      RPG,Adventure   
40833  Action,Horror,Survival Horror,Third-Person Sho...   

                                            game_details   
40829  Single-player,Shared/Split Screen,Downloadable...  \
40830  Multi-player,Co-op,Cross-Platform Multiplayer,...   
40831  Single-player,Downloadable Content,Steam Achie...   
40832  Single-player,Partial Controller Support,Profi...   
40833  Single-player,Steam Achievements,Full controll...   

                                               languages               genre  
40829  English,German,French,Italian,Spanish - Spain,...   Casual,Simulation  
40830                                            English  Indie,RPG,Strategy  
40831  English,French,Italian,German,Spanish - Spain,...              Action  
40832                                            English       Adventure,RPG  
40833              English,French,Italian,German,Spanish    Action,Adventure

**3. Column "all_reviews"**

In [7]:
dataframe_clean.loc[dataframe_clean["all_reviews"].isna()].index

Index([    5,    15,    25,    39,    44,    57,    59,    63,    66,    68,
       ...
       40823, 40824, 40825, 40826, 40827, 40828, 40829, 40830, 40831, 40832],
      dtype='int64', length=12363)

In [8]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["all_reviews"].isna()].index)
dataframe_clean.loc[dataframe_clean["all_reviews"].isna()].index

Index([], dtype='int64')

In [9]:
dataframe_clean.shape

(28471, 9)

**12347 rows less. Sadly, this is the column from where I will be extracting my target data and it is bad idea to fill the missing values as it may affect the model predictions.**

**4. Column "release_date"**

In [10]:
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index([  227,   329,   847,   849,  1039,  1435,  1631,  3280,  3422,  4758,
       ...
       28995, 29243, 29854, 29884, 30014, 30015, 31154, 33952, 34013, 34057],
      dtype='int64', length=132)

In [11]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["release_date"].isna()].index)
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index([], dtype='int64')

In [12]:
dataframe_clean.shape

(28339, 9)

**132 less rows. Insignificant on the chunk of data left.**

**5. Column "developer"**

In [13]:
dataframe_clean.loc[dataframe_clean["developer"].isna()].index

Index([  146,  1101,  1854,  2517,  2712,  3471,  3989,  4722,  4854,  4933,
       ...
       30042, 30246, 30248, 30293, 30608, 31152, 31170, 31231, 34058, 34304],
      dtype='int64', length=155)

In [14]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["developer"].isna()].index)
dataframe_clean.loc[dataframe_clean["developer"].isna()].index

Index([], dtype='int64')

In [15]:
dataframe_clean.shape

(28184, 9)

**155 less rows. Insignificant on the chunk of data left.**

**6. Column "publisher"**

In [16]:
dataframe_clean.loc[dataframe_clean["publisher"].isna()].index

Index([   45,  1061,  1899,  1979,  2422,  2574,  2645,  2711,  2854,  2941,
       ...
       34390, 34529, 34610, 34732, 35581, 36090, 37946, 37947, 37948, 38360],
      dtype='int64', length=2096)

In [17]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["publisher"].isna()].index)
dataframe_clean.loc[dataframe_clean["publisher"].isna()].index

Index([], dtype='int64')

In [18]:
dataframe_clean.shape

(26088, 9)

**2096 less rows. Significant loss, but no way to fill the missing data.**

**7. Column "popular_tags"**

In [19]:
dataframe_clean.loc[dataframe_clean["popular_tags"].isna()].index

Index([ 6027, 10649, 10664, 11206, 11377, 11379, 11814, 11940, 11942, 12508,
       12663, 13206, 13213, 13967, 14104, 14702, 14763, 15580, 15635, 15637,
       15740, 16082, 16139, 16583, 17437, 18580, 19396, 19996, 20268, 21321,
       21642, 22442, 23009, 23050, 23414, 23562, 24129, 25036, 27069, 27845,
       28949, 31218, 31615, 32041, 32045, 34271, 34608],
      dtype='int64')

In [20]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["popular_tags"].isna()].index)
dataframe_clean.loc[dataframe_clean["popular_tags"].isna()].index

Index([], dtype='int64')

In [21]:
dataframe_clean.shape

(26041, 9)

**47 less rows. Insignificant on the chunk of data left.**

**8. Column "game_details"**

In [22]:
dataframe_clean.loc[dataframe_clean["game_details"].isna()].index

Index([  478,   663,   727,   859,   949,  1007,  1052,  1114,  1153,  1305,
       ...
       30623, 30645, 30669, 30758, 30787, 30821, 35821, 36154, 36821, 37456],
      dtype='int64', length=273)

In [23]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["game_details"].isna()].index)
dataframe_clean.loc[dataframe_clean["game_details"].isna()].index

Index([], dtype='int64')

In [24]:
dataframe_clean.shape

(25768, 9)

**273 less rows. Insignificant on the chunk of data left.**

**9. Column "languages"**

In [25]:
dataframe_clean.loc[dataframe_clean["languages"].isna()].index

Index([], dtype='int64')

**No missing data.**

**10. Column "genre"**

In [26]:
dataframe_clean.loc[dataframe_clean["genre"].isna()].index

Index([  528,  1613,  2254,  3639,  4018,  5211,  6127,  6743,  6990,  7138,
        7510,  7836,  8239,  8780, 10677, 11948, 12388, 12987, 13857, 14871,
       14879, 14880, 14884, 15119, 16370, 18200, 18268, 19055, 21998, 22570,
       23695, 23696, 24331, 24884, 26416, 27277, 27402, 29379],
      dtype='int64')

In [27]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["genre"].isna()].index)
dataframe_clean.loc[dataframe_clean["genre"].isna()].index

Index([], dtype='int64')

In [28]:
dataframe_clean.shape

(25730, 9)

**38 less rows. Insignificant on the chunk of data left.**

# 2. Reshaping the dataset

**1. Creating the column "has_setting". It describes wether the game starts the setting or continue it.**

In [29]:
# function looks for numbers in range (2-10) in the column "name" and returns True if there is a number in a string
def check_for_setting(name):
    setting = re.findall(r"\d+", name)
    if setting:
        for i in setting:
            if 2 <= int(i) <= 10:
                return True
    return False

In [30]:
dataframe_clean["has_setting"] = dataframe_clean["name"].apply(check_for_setting)
dataframe_clean = dataframe_clean.set_index("name")

In [31]:
has_setting = dataframe_clean[dataframe_clean["has_setting"]]
has_setting.shape

(2670, 9)

**2669 rows in dataset respond with True for a column "has_setting"**

**2. Changing the datatype in column "release_date" to an int which indicates the year of releas.**

In [32]:
dataframe_clean["release_date"] = pd.to_datetime(dataframe_clean["release_date"], 
                                                 format="%b %d, %Y", errors="coerce").dt.year.astype("Int64")

**Checking if some data was lost in a process and deleting rows where it is absent.**

In [33]:
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index(['Aura Kingdom', 'Technobabylon', 'Age of Wonders Shadow Magic',
       'Puzzle Agent', 'NECROPOLIS: BRUTAL EDITION', 'Red Stone Online',
       'Silent Hunter 5®: Battle of the Atlantic',
       'Sea Dogs: To Each His Own - Pirate Open World RPG', 'Age of Wonders',
       'Idle Racing GO: Clicker Tycoon', 'Muffin Knight',
       'NOBUNAGA'S AMBITION: Souzou (Traditional Chinese version)',
       'DW8XLCE - JAPANESE VOICE OPTION',
       'Nobunaga's Ambition: Souzou - Nobunaga Oda In-Game Face CG',
       'Tetrapulse',
       'Nobunaga's Ambition: Souzou - Series 30th Anniversary Contents',
       'Silent Escape: Induction', 'Apparition', 'Smith and Winston',
       'Aliens vs Predator™ Bughunt Map Pack', 'Depression The Game',
       'Age of Wonders II: The Wizard's Throne', 'Altitude0: Lower & Faster',
       'Turba', 'Call of Duty®: Advanced Warfare - Season Pass',
       'NIGHTSTAR: Rogue Wings', 'Cirque du Soleil',
       '3DMark Time Spy benchmark', 'Darkest of Days', 'Blue

In [34]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["release_date"].isna()].index)
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index([], dtype='object', name='name')

In [35]:
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date              developer   
name                                                       
Die In The Dark              2018          Side B Gaming  \
Chasing the Stars            2019            Ertal Games   
Fruit Tower Defense          2018  Fratboy Entertainment   
Nanoui                       2018               Phoesion   
Resident Evil 4              2023        CAPCOM CO., LTD   

                                                       publisher   
name                                                               
Die In The Dark                      Side B Gaming,Side B Gaming  \
Chasing the Stars                        Ertal Games,Ertal Games   
Fruit Tower Defense  Fratboy Entertainment,Fratboy Entertainment   
Nanoui                                         Phoesion,Phoesion   
Resident Evil 4                                  CAPCOM CO., LTD   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 languages   
name                                                         
Die In The Dark                                    English  \
Chasing the Stars                                  English   
Fruit Tower Defense                                English   
Nanoui                                             English   
Resident Evil 4      English,French,Italian,German,Spanish   

                                                 genre  has_setting  
name                                                                 
Die In The Dark             Adventure,Indie,Simulation        False  
Chasing the Stars                                Indie        False  
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False  
Nanoui               Action,Adventure,Casual,Indie,RPG        False  
Resident Evil 4                       Action,Adventure         True

**33 rows were deleted in a process.**

**3. Creating the column "published_by_developer". The column will have a boolean type of data regarding the data in columns "publisher" and "developer".**

**removing the duplicated strings in column "publisher".**

In [36]:
# function spilts the string and returns unuiqe part of it
def remove_duplicates(string):
    publishers = string.split(",")
    unique_publishers = list(set(publishers))
    return ",".join(unique_publishers)

In [37]:
dataframe_clean["publisher"] = dataframe_clean["publisher"].apply(remove_duplicates)

In [38]:
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date              developer   
name                                                       
Die In The Dark              2018          Side B Gaming  \
Chasing the Stars            2019            Ertal Games   
Fruit Tower Defense          2018  Fratboy Entertainment   
Nanoui                       2018               Phoesion   
Resident Evil 4              2023        CAPCOM CO., LTD   

                                 publisher   
name                                         
Die In The Dark              Side B Gaming  \
Chasing the Stars              Ertal Games   
Fruit Tower Defense  Fratboy Entertainment   
Nanoui                            Phoesion   
Resident Evil 4             LTD,CAPCOM CO.   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 languages   
name                                                         
Die In The Dark                                    English  \
Chasing the Stars                                  English   
Fruit Tower Defense                                English   
Nanoui                                             English   
Resident Evil 4      English,French,Italian,German,Spanish   

                                                 genre  has_setting  
name                                                                 
Die In The Dark             Adventure,Indie,Simulation        False  
Chasing the Stars                                Indie        False  
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False  
Nanoui               Action,Adventure,Casual,Indie,RPG        False  
Resident Evil 4                       Action,Adventure         True

**Creating the column "published_by_developer" and deliting columns "developer" and "publisher".**

In [39]:
dataframe_clean["published_by_developer"] = dataframe_clean["developer"] == dataframe_clean["publisher"]

In [40]:
published_by_developer = dataframe_clean[dataframe_clean["published_by_developer"]]
dataframe_clean = dataframe_clean.drop(["developer", "publisher"], axis=1)
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date   
name                                
Die In The Dark              2018  \
Chasing the Stars            2019   
Fruit Tower Defense          2018   
Nanoui                       2018   
Resident Evil 4              2023   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 languages   
name                                                         
Die In The Dark                                    English  \
Chasing the Stars                                  English   
Fruit Tower Defense                                English   
Nanoui                                             English   
Resident Evil 4      English,French,Italian,German,Spanish   

                                                 genre  has_setting   
name                                                                  
Die In The Dark             Adventure,Indie,Simulation        False  \
Chasing the Stars                                Indie        False   
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False   
Nanoui               Action,Adventure,Casual,Indie,RPG        False   
Resident Evil 4                       Action,Adventure         True   

                     published_by_developer  
name                                         
Die In The Dark                        True  
Chasing the Stars                      True  
Fruit Tower Defense                    True  
Nanoui                                 True  
Resident Evil 4                       False

**4. Creating the column "multiple_languages". The column will have a boolean type of data regarding the data in column "languages". Drop the cloumn "languages" after data extraction.**

In [41]:
dataframe_clean["multiple_languages"] = dataframe_clean["languages"].str.split(",").apply(lambda x: len(x) > 1)
dataframe_clean = dataframe_clean.drop(["languages"], axis=1)
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date   
name                                
Die In The Dark              2018  \
Chasing the Stars            2019   
Fruit Tower Defense          2018   
Nanoui                       2018   
Resident Evil 4              2023   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 genre  has_setting   
name                                                                  
Die In The Dark             Adventure,Indie,Simulation        False  \
Chasing the Stars                                Indie        False   
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False   
Nanoui               Action,Adventure,Casual,Indie,RPG        False   
Resident Evil 4                       Action,Adventure         True   

                     published_by_developer  multiple_languages  
name                                                             
Die In The Dark                        True               False  
Chasing the Stars                      True               False  
Fruit Tower Defense                    True               False  
Nanoui                                 True               False  
Resident Evil 4                       False                True

In [42]:
multiple_languages = dataframe_clean[dataframe_clean["multiple_languages"]]
multiple_languages.shape

(13611, 8)

**13610 rows respond with True for column "multiple_languages".**

**5. Creating multiple columns regarding unique tags, details and genres. They will store boolean data type regarding the string they were extracted from.**

**Listing and counting the tags from column "popular_tags".**

In [43]:
all_tags = ",".join(dataframe_clean["popular_tags"])
tag_list = all_tags.split(',')
tag_counts = Counter(tag_list)
most_popular_tags = tag_counts.most_common()
print("Most popular tags:")
for tag, count in most_popular_tags:
    if count > 300:
        print(f"{tag}: {count}")

Most popular tags:
Indie: 16669
Action: 11794
Adventure: 9945
Casual: 8822
Simulation: 6183
Singleplayer: 5976
Strategy: 5794
Early Access: 4959
RPG: 4770
Great Soundtrack: 2824
Multiplayer: 2752
2D: 2695
Atmospheric: 2673
Puzzle: 2581
VR: 2043
Free to Play: 1898
Story Rich: 1872
Difficult: 1823
Violent: 1721
Horror: 1694
Anime: 1662
Pixel Graphics: 1552
Funny: 1521
Shooter: 1499
Platformer: 1491
Gore: 1468
Open World: 1455
First-Person: 1454
Sci-fi: 1447
Female Protagonist: 1439
Fantasy: 1424
Co-op: 1417
Sports: 1302
Retro: 1295
FPS: 1210
Arcade: 1200
Nudity: 1123
Survival: 1118
Family Friendly: 1088
Racing: 1058
Visual Novel: 1016
Comedy: 1005
Sandbox: 1002
Classic: 954
Cute: 948
Point & Click: 936
Massively Multiplayer: 931
Sexual Content: 912
Turn-Based: 852
Exploration: 850
Masterpiece: 849
Replay Value: 776
Space: 768
Psychological Horror: 747
Third Person: 732
Relaxing: 730
Local Multiplayer: 638
Mystery: 615
Colorful: 611
Tactical: 602
Zombies: 589
Controller: 589
Fast-Paced: 5

**Creating new columns according to uniqe tags.**

In [44]:
the_most_popular_tags = [(tag, count) for tag, count in tag_counts.items() if count > 300]
for tag, count in the_most_popular_tags:
    dataframe_clean[tag] = dataframe_clean["popular_tags"].apply(lambda x: tag in x)

In [45]:
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date   
name                                
Die In The Dark              2018  \
Chasing the Stars            2019   
Fruit Tower Defense          2018   
Nanoui                       2018   
Resident Evil 4              2023   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 genre  has_setting   
name                                                                  
Die In The Dark             Adventure,Indie,Simulation        False  \
Chasing the Stars                                Indie        False   
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False   
Nanoui               Action,Adventure,Casual,Indie,RPG        False   
Resident Evil 4                       Action,Adventure         True   

                     published_by_developer  multiple_languages    FPS   Gore   
name                                                                            
Die In The Dark                        True               False  False   True  \
Chasing the Stars                      True               False  False  False   
Fruit Tower Defense                    True               False  False  False   
Nanoui                                 True               False  False  False   
Resident Evil 4                       False                True  False   True   

                     ...  Relaxing  Replay Value  Education   
name                 ...                                      
Die In The Dark      ...     False         False      False  \
Chasing the Stars    ...     False         False      False   
Fruit Tower Defense  ...     False         False      False   
Nanoui               ...     False         False      False   
Resident Evil 4      ...     False         False      False   

                     Design & Illustration  Procedural Generation  Music   
name                                                                       
Die In The Dark                      False                  False  False  \
Chasing the Stars                    False                  False  False   
Fruit Tower Defense                  False                  False  False   
Nanoui                               False                  False  False   
Resident Evil 4                      False                  False  False   

                     Shoot 'Em Up  RPGMaker  Hidden Object  Minimalist  
name                                                                    
Die In The Dark             False     False          Fa

**Listing and counting the tags from column "game_details".**

In [46]:
all_details = ",".join(dataframe_clean["game_details"])
details_list = all_details.split(',')
details_counts = Counter(details_list)
most_popular_details = details_counts.most_common()
print("Most popular details:")
for detail, count in most_popular_details:
    if count > 300:
        print(f"{detail}: {count}")

Most popular details:
Single-player: 24018
Steam Achievements: 14120
Steam Trading Cards: 8425
Steam Cloud: 8188
Full controller support: 6288
Downloadable Content: 5357
Multi-player: 5289
Partial Controller Support: 4443
Profile Features Limited 
									: 4304
Steam Leaderboards: 3758
Online Multi-Player: 2771
Co-op: 2463
Shared/Split Screen: 2130
Stats: 1848
Steam Workshop: 1461
Local Multi-Player: 1434
Online Co-op: 1301
Cross-Platform Multiplayer: 1259
Includes level editor: 1133
Local Co-op: 951
In-App Purchases: 869
Captions available: 785
Steam is learning about this game 
									: 666
MMO: 536


**Creating new columns according to uniqe details.**

In [47]:
the_most_popular_details = [(detail, count) for detail, count in details_counts.items() if count > 500]
for detail, count in the_most_popular_details:
    new_column_name = f"detail_{detail}"
    dataframe_clean[new_column_name] = dataframe_clean["game_details"].apply(lambda x: detail in x)

In [48]:
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date   
name                                
Die In The Dark              2018  \
Chasing the Stars            2019   
Fruit Tower Defense          2018   
Nanoui                       2018   
Resident Evil 4              2023   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 genre  has_setting   
name                                                                  
Die In The Dark             Adventure,Indie,Simulation        False  \
Chasing the Stars                                Indie        False   
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False   
Nanoui               Action,Adventure,Casual,Indie,RPG        False   
Resident Evil 4                       Action,Adventure         True   

                     published_by_developer  multiple_languages    FPS   Gore   
name                                                                            
Die In The Dark                        True               False  False   True  \
Chasing the Stars                      True               False  False  False   
Fruit Tower Defense                    True               False  False  False   
Nanoui                                 True               False  False  False   
Resident Evil 4                       False                True  False   True   

                     ...  detail_Local Co-op  detail_Shared/Split Screen   
name                 ...                                                   
Die In The Dark      ...               False                       False  \
Chasing the Stars    ...               False                       False   
Fruit Tower Defense  ...               False                       False   
Nanoui               ...                True                        True   
Resident Evil 4      ...               False                       False   

                     detail_Steam Leaderboards  detail_Local Multi-Player   
name                                                                        
Die In The Dark                          False                      False  \
Chasing the Stars                        False                      False   
Fruit Tower Defense                      False                      False   
Nanoui                                   False                       True   
Resident Evil 4                          False                      False   

                     detail_Captions available  detail_Includes level editor   
name                   

**Listing and counting the tags from column "genre".**

In [49]:
all_genres = ",".join(dataframe_clean["genre"])
genre_list = all_genres.split(",")
genre_counts = Counter(genre_list)
most_popular_genres = [(genre, count) for genre, count in genre_counts.items() if count > 300]
print("Most popular genres:")
for genre, count in most_popular_genres:
    print(f"{genre}: {count}")

Most popular genres:
Action: 11065
Adventure: 8827
Massively Multiplayer: 809
Strategy: 5258
Free to Play: 1700
RPG: 4417
Indie: 15950
Early Access: 2193
Simulation: 5705
Racing: 994
Casual: 7982
Sports: 1224


**Creating new columns according to uniqe genres.**

In [50]:
the_most_popular_genres = [(genre, count) for genre, count in genre_counts.items() if count > 500]
for genre, count in the_most_popular_genres:
    new_column_name = f"genre_{genre}"
    dataframe_clean[new_column_name] = dataframe_clean["genre"].apply(lambda x: genre in x)

In [51]:
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date   
name                                
Die In The Dark              2018  \
Chasing the Stars            2019   
Fruit Tower Defense          2018   
Nanoui                       2018   
Resident Evil 4              2023   

                                                          popular_tags   
name                                                                     
Die In The Dark                Adventure,Gore,Indie,Violent,Simulation  \
Chasing the Stars    Nudity,Sexual Content,Indie,Steampunk,Story Ri...   
Fruit Tower Defense      Strategy,Adventure,Indie,Casual,Tower Defense   
Nanoui               Casual,RPG,Action,Adventure,Indie,Violent,Co-o...   
Resident Evil 4      Action,Horror,Survival Horror,Third-Person Sho...   

                                                          game_details   
name                                                                     
Die In The Dark      Single-player,Profile Features Limited \r\n\t\...  \
Chasing the Stars    Single-player,Steam Achievements,Profile Featu...   
Fruit Tower Defense  Single-player,Steam Achievements,Profile Featu...   
Nanoui               Single-player,Multi-player,Online Multi-Player...   
Resident Evil 4      Single-player,Steam Achievements,Full controll...   

                                                 genre  has_setting   
name                                                                  
Die In The Dark             Adventure,Indie,Simulation        False  \
Chasing the Stars                                Indie        False   
Fruit Tower Defense    Adventure,Casual,Indie,Strategy        False   
Nanoui               Action,Adventure,Casual,Indie,RPG        False   
Resident Evil 4                       Action,Adventure         True   

                     published_by_developer  multiple_languages    FPS   Gore   
name                                                                            
Die In The Dark                        True               False  False   True  \
Chasing the Stars                      True               False  False  False   
Fruit Tower Defense                    True               False  False  False   
Nanoui                                 True               False  False  False   
Resident Evil 4                       False                True  False   True   

                     ...  genre_Massively Multiplayer  genre_Strategy   
name                 ...                                                
Die In The Dark      ...                        False           False  \
Chasing the Stars    ...                        False           False   
Fruit Tower Defense  ...                        False            True   
Nanoui               ...                        False           False   
Resident Evil 4      ...                        False           False   

                     genre_Free to Play  genre_RPG  genre_Indie   
name                                                              
Die In The Dark                   False      False         True  \
Chasing the Stars                 False      False         True   
Fruit Tower Defense               False      False         True   
Nanoui                            False       True         True   
Resident Evil 4                   False      False        False   

                     genre_Early Access  genre_Simulation  genre_Racing   
name                                                                      
Die In The Dark                   False     

**Deleting the columns "genre", "popular_tags" and "game_details" as I no longer need them.**

In [52]:
dataframe_clean = dataframe_clean.drop(["genre", "popular_tags", "game_details"], axis=1)
dataframe_clean.tail()

all_reviews   
name                                                                     
Die In The Dark      1 user reviews,- Need more user reviews to gen...  \
Chasing the Stars    Mostly Positive,(10),- 70% of the 10 user revi...   
Fruit Tower Defense  4 user reviews,- Need more user reviews to gen...   
Nanoui               Mostly Negative,(14),- 28% of the 14 user revi...   
Resident Evil 4                Mixed,(16),- 54% of the 16 user reviews   

                     release_date  has_setting  published_by_developer   
name                                                                     
Die In The Dark              2018        False                    True  \
Chasing the Stars            2019        False                    True   
Fruit Tower Defense          2018        False                    True   
Nanoui                       2018        False                    True   
Resident Evil 4              2023         True                   False   

                     multiple_languages    FPS   Gore  Action  Shooter   
name                                                                     
Die In The Dark                   False  False   True   False    False  \
Chasing the Stars                 False  False  False   False    False   
Fruit Tower Defense               False  False  False   False    False   
Nanoui                            False  False  False    True    False   
Resident Evil 4                    True  False   True    True     True   

                     First-Person  ...  genre_Massively Multiplayer   
name                               ...                                
Die In The Dark             False  ...                        False  \
Chasing the Stars           False  ...                        False   
Fruit Tower Defense         False  ...                        False   
Nanoui                      False  ...                        False   
Resident Evil 4             False  ...                        False   

                     genre_Strategy  genre_Free to Play  genre_RPG   
name                                                                 
Die In The Dark               False               False      False  \
Chasing the Stars             False               False      False   
Fruit Tower Defense            True               False      False   
Nanoui                        False               False       True   
Resident Evil 4               False               False      False   

                     genre_Indie  genre_Early Access  genre_Simulation   
name                                                                     
Die In The Dark             True               False              True  \
Chasing the Stars           True               False             False   
Fruit Tower Defense         True               False             False   
Nanoui                      True               False             False   
Resident Evil 4            False               False             False   

                     genre_Racing  genre_Casual  genre_Sports  
name                                                           
Die In The Dark             False         False         False  
Chasing the Stars           False         False         False  
Fruit Tower Defense         False          True         False  
Nanoui                      False          True         False  
Resident Evil 4             False         False         False  

[5 rows x 149 columns]

In [53]:
dataframe_clean.shape

(25698, 149)

**6. Creating columns "total_reviews" and "positive_reviews_share" regarding the column "all_reviews". They will contain int and float data type accordingly.**

**Checking if all rows have a sufficient string to extract data. Deleting those which do not have.**

In [54]:
indices = dataframe_clean[dataframe_clean["all_reviews"].str.contains("Need more", na=False)].index
dataframe_clean = dataframe_clean.drop(indices)

In [55]:
dataframe_clean.shape

(16345, 149)

**9353 rows were deleted as they lacked required data.**

In [56]:
# function extracts first and second number from a string and returns them as int and float data type accordingly
def extract_numbers(string):
    total_reviews = None
    positive_reviews_share = None
    match_total = re.search(r"\(([\d,]+)\)", string)
    match_positive = re.search(r"(\d+)%", string)
    if match_total:
        total_reviews = int(match_total.group(1).replace(",", ""))
    if match_positive:
        positive_reviews_share = float(match_positive.group(1)) / 100
    return total_reviews, positive_reviews_share

In [57]:
dataframe_clean[["total_reviews", 
                 "positive_reviews_share"]] = dataframe_clean["all_reviews"].apply(lambda x: pd.Series(extract_numbers(x)))

**Deleting the column "all_reviews" as data was exctracted already.**

In [58]:
dataframe_clean = dataframe_clean.drop(["all_reviews"], axis=1)

**Selecting a required row for the prediction.**

In [59]:
dataframe_clean = dataframe_clean.loc[["Resident Evil 4"]]

**Checking on how dataset looks like now.**

In [60]:
dataframe_clean

release_date  has_setting  published_by_developer   
name                                                                 
Resident Evil 4          2023         True                   False  \

                 multiple_languages    FPS  Gore  Action  Shooter   
name                                                                
Resident Evil 4                True  False  True    True     True  \

                 First-Person  Great Soundtrack  ...  genre_Free to Play   
name                                             ...                       
Resident Evil 4         False             False  ...               False  \

                 genre_RPG  genre_Indie  genre_Early Access  genre_Simulation   
name                                                                            
Resident Evil 4      False        False               False             False  \

                 genre_Racing  genre_Casual  genre_Sports  total_reviews   
name                                                                       
Resident Evil 4         False         False         False           16.0  \

                 positive_reviews_share  
name                                     
Resident Evil 4                    0.54  

[1 rows x 150 columns]

# 4. Saving the reshaped dataset

In [61]:
dataframe_clean.to_csv("Datasets/prediction_1.csv", index=False)